File: src/main/java/pl/allegro/experiments/chi/chiserver/application/interactions/v1/InteractionTypeAdapter.java
package pl.allegro.experiments.chi.chiserver.application.interactions.v1;

import com.google.gson.*;
import pl.allegro.experiments.chi.chiserver.domain.interactions.Interaction;

import java.lang.reflect.Type;
import java.time.Instant;

import static pl.allegro.experiments.chi.chiserver.application.JsonObjectGetters.getAsBooleanOrNull;
import static pl.allegro.experiments.chi.chiserver.application.JsonObjectGetters.getAsStringOrNull;

public class InteractionTypeAdapter implements JsonDeserializer<Interaction> {

    @Override
    public Interaction deserialize(
            JsonElement json,
            Type typeOfT,
            JsonDeserializationContext context) throws JsonParseException {
        JsonObject jsonObj = json.getAsJsonObject();

        return new Interaction(
                getAsStringOrNull("userId", jsonObj),
                getAsStringOrNull("userCmId", jsonObj),
                jsonObj.get("experimentId").getAsString(),
                jsonObj.get("variantName").getAsString(),
                getAsBooleanOrNull("internal", jsonObj),
                getAsStringOrNull("deviceClass", jsonObj),
                Instant.parse(jsonObj.get("interactionDate").getAsString()),
                getAsStringOrNull("appId", jsonObj)
        );
    }

}

File: src/main/kotlin/pl/allegro/experiments/chi/chiserver/application/interactions/v1/InteractionTypeAdapter.kt
package pl.allegro.experiments.chi.chiserver.application.interactions.v1

import com.google.gson.*
import pl.allegro.experiments.chi.chiserver.domain.interactions.Interaction
import java.lang.reflect.Type
import java.time.Instant

class InteractionTypeAdapter: JsonDeserializer<Interaction> {
    override fun deserialize(
            json: JsonElement,
            typeOfT: Type?,
            context: JsonDeserializationContext?): Interaction {
        val jsonObj = json.asJsonObject
        return Interaction(
                jsonObj.get("userId")?.asString,
                jsonObj.get("userCmId")?.asString,
                jsonObj.get("experimentId").asString,
                jsonObj.get("variantName").asString,
                jsonObj.get("internal")?.asBoolean,
                jsonObj.get("deviceClass")?.asString,
                Instant.parse(jsonObj.get("interactionDate").asString),
                jsonObj.get("appId")?.asString
        )
    }
}